resources/assets/coffee/react/_components/bbcode-editor.coffee

```diff
@@ -0,0 +1,200 @@
+###
+#    Copyright 2015-2017 ppy Pty. Ltd.
+#
+#    This file is part of osu!web. osu!web is distributed with the hope of
+#    attracting more community contributions to the core ecosystem of osu!.
+#
+#    osu!web is free software: you can redistribute it and/or modify
+#    it under the terms of the Affero GNU General Public License version 3
+#    as published by the Free Software Foundation.
+#
+#    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+#    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#    See the GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+
+{button, div, em, i, label, option, select, span, strong} = ReactDOMFactories
+el = React.createElement
+
+class @BBCodeEditor extends React.Component
+  constructor: (props) ->
+    super props
+
+    @state =
+      disabled: @props.disabled
+      rawValue: @props.rawValue
+
+
+  componentDidMount: =>
+    if @props.selection?.range
+      @refs.body.selectionStart = @props.selection.range[0]
+      @refs.body.selectionEnd = @props.selection.range[1]
+
+    @refs.body.focus()
+
+
+  componentWillUnmount: =>
+    @props.onSelectionUpdate && @props.onSelectionUpdate
+      range: [@refs.body.selectionStart, @refs.body.selectionEnd]
+
+
+  onKeyDown: (e) =>
+    e.keyCode == 27 && @_cancel()
+
+
+  onInput: (_e) =>
+    @setState rawValue: @refs.body.value
+
+
+  _cancel: =>
+    @refs.body.value = @props.rawValue
+    @setState rawValue: @props.rawValue
+    @props.onChange(type: 'cancel', value: @props.rawValue) if @props.onChange
+
+
+  _reset: =>
+    @refs.body.value = @props.rawValue
+    @setState dirty: false, rawValue: @props.rawValue
+    @props.onChange(type: 'reset', value: @props.rawValue) if @props.onChange
+
+
+  _save: =>
+    if @props.onChange
+      @props.onChange
+        type: 'save'
+        value: @state.rawValue
+        hasChanged: @state.rawValue != @props.rawValue
+
+
+  render: ->
+    el 'form', className: 'post-editor',
+      el 'textarea',
+        className: 'post-editor__textarea'
+        name: 'body'
+        value: @state.rawValue
```

Comment:
Looks like uncontrolled component is better for this and create `rawValue` function which returns `@body.value` instead of manual synchronization between state and actual dom.